---
title: "assignment"
author: "Audran Feuvrier"
date: "2024-03-01"
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(AnnotationHub) # to fetch genomes, annotations, etc.
  library(ensembldb)     # to manipulate ensembl gene annotations
  library(GenomicRanges) # data structure for genomic ranges 
  library(epiwraps)      # visualization
  library(ggplot2)
})
```


```{r}
ah <- AnnotationHub()

```

The mouse (Mus Musculus) EnsDb object, version 102, genome build GRCm38

```{r}
q <- query(ah, c(species="Mus musculus", dataprovider="Ensembl", rdataclass="EnsDb", genome="GRCm38", 102 )) 
q
#AH89211 

```
We can find a record from Ensembl of Mus musculus version 102 with the name AH89211 in a EnsDb format.

```{r}
ensdb=ah[["AH89211"]]
```


The mouse genome sequence ( dna_sm ) in TwoBit/2bit format for GRCm38

```{r}
q2 <- query(ah, c(species="Mus musculus", dataprovider="Ensembl", rdataclass="TwoBitFile", genome="GRCm38", 102 )) 
q2
```

The dna_sm file we are looking for has the name "AH88477".
```{r}
twobit_mouse = ah[["AH88477"]]

```


The drosophila melanogaster genome sequence ( dna_sm ) in TwoBit/2bit format for BDGP6
```{r}
q3 <- query(ah, c(species="Drosophila melanogaster", dataprovider="Ensembl", rdataclass="TwoBitFile", genome="BDGP6")) 
q3
```

Version 32 seems to be the most recent record for Drosophila melanogaster dna_sm two bit file. The corresponding record name is "AH106160"

```{r}
twobit_droso = ah[["AH106160"]]

```


```{r}
g=genes(ensdb) #retrieve all genes in record AH89211
head(g)
```

How many different protein coding gene ID are there?
```{r}
pcg=g[which(g$gene_biotype=="protein_coding")] #keep only the protein coding genes within g

length(pcg$gene_id)

length(unique(pcg$gene_id)) #I'm using unique() to make sure there aren't any duplicates even if I'm not expecting to see any
```

How many different gene symbols?
```{r}
length(pcg$symbol)
length(unique(pcg$symbol))
```
There isn't as many gene IDs as symbols! Certain protein coding gene seem to have the same symbol.


Plot the distribution of the (spliced) length of protein-coding transcripts

    (tip: this will require you to extract exons of protein-coding transcripts from the database, and split them by transcript, before summing the width of the exons of each transcript)

```{r}
tx=transcripts(ensdb) #retrieve all transcripts

ex=exonsBy(ensdb)[which(tx$tx_biotype=="protein_coding")] #keep only the protein coding transcripts and retrieve their exons by transcripts

```


```{r}
to_plot=sum(width(ranges(ex))) #sum the width of the ranges of all exons per protein coding transcripts
ggplot(data.frame(length=to_plot), aes(length)) + geom_histogram() + scale_x_log10()
```


