---
title: "assignment"
author: "Audran Feuvrier"
date: "2024-04-22"
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(epiwraps)
  library(AnnotationHub)
  library(MotifDb)
  library(universalmotif)
  library(ensembldb)
  library(ggplot2)
  library(magick)
})
ah <- AnnotationHub(localHub=TRUE)

```

In the same dataset of ATAC on chr19, plot 1) the insertion (i.e. ‘cuts’) profile of nucleosome-free fragments and 2) the centers of nucleosome-containing fragments, around the high-confidence motifs of two factors.
You can choose your own factors of interest, or for instance use KLF4 and MAZ.

Get the motifs of two factors:
```{r}
motif_KLF4 <- MotifDb::query(MotifDb, c("KLF4","Mus"))[[1]] #get the HOCOMOCO mouse motif 
motif_MAZ <- MotifDb::query(MotifDb, c("MAZ","Mus"))[[1]] #get the HOCOMOCO mouse motif 
motif2_KLF4 <- convert_motifs(motif_KLF4, class="TFBSTools-PFMatrix") #put it in the right format
motif2_MAZ <- convert_motifs(motif_MAZ, class="TFBSTools-PFMatrix") #put it in the right format

```
Get the corresponding genome and the right chromosome (chr19)
```{r}
genome <- ah[["AH68356"]]
# get the sequence for chr19:
chr19 <- import(genome)["19"] 
```

```{r}
# find motif matches across chr19 for both factors
moi <- motifmatchr::matchMotifs(motif2_KLF4, chr19, out="positions", p.cutoff=1e-5)[[1]]
# convert to GRanges
moi <- as(setNames(moi,names(chr19)), "GRanges")

# find motif matches across chr19
moi2 <- motifmatchr::matchMotifs(motif2_MAZ, chr19, out="positions", p.cutoff=1e-5)[[1]]
# convert to GRanges
moi2 <- as(setNames(moi2,names(chr19)), "GRanges")
```


```{r, fig.width=8, fig.height=4}
# we prepare the list of tracks
tracks <- c("NF coverage"="NF_cov.bw", "Mono centers"="mono_centers.bw")

# extract signals around the motif occurences
# we zoom in to 300bp around the motif centers, in windows of 5bp
sm <- signal2Matrix(tracks, moi, w=5, extend=300)


#background normalization
# this we can do for instance using:
nf <- getNormFactors(tracks, useSeqLevels="19", nwind=5000L)
# then we apply the normalization factors:
sm <- renormalizeSignalMatrices(sm, scaleFactors = nf)

plotEnrichedHeatmaps(sm, trim=0.95, minRowVal = 12, colors = c("white","darkred"))
```


```{r}

# extract signals around the motif occurences
# we zoom in to 300bp around the motif centers, in windows of 5bp
sm2 <- signal2Matrix(tracks, moi2, w=5, extend=300)

sm2 <- renormalizeSignalMatrices(sm2, scaleFactors = nf)
plotEnrichedHeatmaps(sm2, trim=0.95, minRowVal = 12, colors = c("white","darkred"))

```



